File: 卡码/代码随想录/Go/回溯/02-分割/LeetCode_0093.go
// 93. 复原 IP 地址
package main

import "strings"

func isValid(s string,start,end int) bool{
	if start>end||s[start]=='0'&&start!=end {
		return false
	}
	num:=0
	for i:=start;i<=end;i++ {
		digit:=int(s[i]-'0')
		num=num*10+digit
		if num>255 {
			return false
		}
	}
	return true
}

func restoreIpAddresses(s string) []string {
	var res=make([]string,0)
	var path=make([]string,0)
	var backtracking func(s string,index,dotNum int)
	backtracking=func(s string,index,dotNum int){
		if index==len(s)&&dotNum==4 {
			res=append(res,strings.Join(path,"."))
			return
		}
		for i:=index;i<len(s)&&i-index<3;i++ {
			if isValid(s,index,i) {
				path=append(path,s[index:i+1])
				backtracking(s,i+1,dotNum+1)
				path=path[:len(path)-1]
			}else{
				break
			}
		}
	}
	backtracking(s,0,0)
	return res
}

File: 卡码/代码随想录/JS/回溯/02-分割/LeetCode_0093.js
//93. 复原 IP 地址
const isValid=function(s,start,end){
  if(start>end||s[start]=="0"&&start!=end) return false;
  let num=0;
  const digit=parseInt(s.slice(start,end+1));
  if(digit>255) return false;
  return true;
}
const restoreIpAddresses = function(s) {
  const res=new Array();
  const path=new Array();
  const backtracking=function(s,index,dotNum){
    if(index==s.length&&dotNum==4){
      res.push(path.join("."));
      return;
    }
    for(let i=index;i<s.length&&i-index<3;i++){
      if(isValid(s,index,i)){
        path.push(s.slice(index,i+1));
        backtracking(s,i+1,dotNum+1);
        path.pop();
      }else break;
    }
  }
  backtracking(s,0,0);
  return res;
};